### ISSUE TITLE
withStatement & withBind aren't quite useful

### ISSUE THREAD

**Issue by ip1981** [nurpax/sqlite-simple#50]
`withStatement` & `withBind` aren't quite useful if you want to do a bunch of `INSERTS`.

I think I can use `nextRow` and discard the result, but this require specific type, e. g. `IO (Maybe [Int])`.

It would be great if `Base.step` were exported somehow through the `Statement Base.Statement` wrapper.

I _could_ use `Base.step`, If I could get Base.Statement from Statement. But `Statement` constructer if not exported either.

**Comment by nurpax** [comment]
I recently added executeMany which can be used for multi-inserts.  Does that fit the bill or do you think we need a similar API to use with statements?

**Comment by ip1981** [comment]
Yeah... it's close, but not enough. In one of my use cases, number of rows is unknown: https://github.com/ip1981/sproxy2/blob/a3b5b7ac059e473e10749591c11aee97482b88d1/src/Sproxy/Server/DB.hs#L137-L139. It smells like conduit with PostgreSQL query as a producer :)

**Comment by nurpax** [comment]
Hi @ip1981 Sorry for taking so long to reply on this.

I wonder if it'd be enough to add a FromRow instance for ().  So you could do

```
submit st v = SQLite.withBind st v $ void (SQLite.nextRow st :: IO (Maybe ()))
```

I think this would be fairly orthogonal to existing features.  E.g., there's also a ToRow instance for () that converts the unit value to an empty list.

If we did this, the returned value from the statement would be error checked against an empty list.  E.g., you wouldn't be able to use a unit nextRow on a SELECT statement.

Does this sound reasonable?  I'll make a PR and send a review request.

**Comment by nurpax** [comment]
Actually, that didn't work out so well.  `nextRow` always returns `Nothing` with inserts (since there are no rows to return).  

Existing code:
```
nextRow :: (FromRow r) => Statement -> IO (Maybe r)
nextRow = nextRowWith fromRow

nextRowWith :: RowParser r -> Statement -> IO (Maybe r)
nextRowWith fromRow_ (Statement stmt) = do
  statRes <- Base.step stmt
  case statRes of
    Base.Row -> do
      rowRes <- Base.columns stmt
      let nCols = length rowRes
      row <- convertRow fromRow_ rowRes nCols
      return $ Just row
    Base.Done -> return Nothing
```

I guess I could either just expose `step` as is, or name it something like `nextRow_` with an IO () return type.

**Comment by nurpax** [comment]
I played with the unit FromRow instance but it might be kind of a bad idea.

You can see the code here: https://github.com/nurpax/sqlite-simple/compare/unit-nextrow

With this feature, it still wouldn't be obvious how to use nextRow with INSERTs (you'd need come up with the idea to use FromRow () and/or read the docs quite a bit).  I think a more explicit API for inserts would be better.  I'm not sure I like my above nextRow_ proposal either.

**Comment by julmb** [comment]
Sorry for reviving a very old thread, but I would actually be interested in having something like this.

I am trying to `INSERT` values from a `ConduitT` (but really, any IO action producing values has a similar shape). The `executeMany` function forces me to either have all values in memory, or do some lazy IO hackery. It would be nice if there was something more explicit.

The changes from 406d9b2 would actually work pretty well for my case. I am slightly worried about parsing overhead, but since the `INSERT` use case produces no result rows, `step` will return `Done`, so the parser never gets invoked, if I understand this correctly. Having a `FromRow ()` instance that checks that there are zero columns also sounds pretty useful in general.

Alternatively, exporting `nextRowWith` so that I can supply my own no-op or failure parser (without orphan instances on `()` or such) would also work.

**Comment by julmb** [comment]
Actually, for the `INSERT` statement, we do not want to check for the condition "all returned rows have zero columns" but instead we want to check "there are no returned rows". So maybe it would actually make even more sense to have a `FromRow Void` instance whose parser always fails. Then instead instead of a monadic bind failure `Nothing <- nextRow stmt :: IO (Maybe ())`, we can handle the `Just` case via `maybe () absurd` and the failure happens inside `nextRow` instead of outside. Then again, this is pretty close to the `nextRow_` proposal which would also package the "no rows returned" check.
